### PR TITLE
Utilize numpy to reduce calculation time

### PR DESCRIPTION
**Pull request by kadirhas**
Change-Id: I6e421a1c2524a3d8f8875121a1a6d2ed832c3150

#### What does this implement/fix?

I've noticed that the current Implementation uses lists. By replacing them with numpy arrays and using built in numpy functions such as any, I reduced the calculation time of d_star_lite. The other thing that I've noticed, once after initializing class, if the user wants to add new spoofed objects and calculate a new path, it was not using the prior information that is generated. So I modified the initialization process a bit to use the prior information.
#### CheckList
- [ ] Did you add an unittest for your new example or defect fix?
- [ ] Did you add documents for your new example?
- [ ] All CIs are green? (You can check it after submitting)

### CODE REVIEW

**Comment by kadirhas** [comment]
The styling warnings are caused by W503, which says that it is [going to change](https://www.flake8rules.com/rules/W503.html).

File: PathPlanning/DStarLite/d_star_lite.py [hunk 1]
```diff
@@ -167,23 +187,25 @@ def compare_keys(self, key_pair1: tuple[float, float],
...
-                                 self.calculate_key(self.start))) or \
-                self.rhs[self.start.x][self.start.y] != \
-                self.g[self.start.x][self.start.y]:
+        while (
+            len(self.U) > 0
+            and self.compare_keys(self.U[0][1], self.calculate_key(self.start))
+        ) or (
+            self.rhs[self.start.x, self.start.y]
+            != self.g[self.start.x, self.start.y]
+        ).any():
```
Comment (AtsushiSakai) [inline PathPlanning/DStarLite/d_star_lite.py:196]: Not this PR topic, but I think each while condition is challenging to read, so is it possible to introduce boolean variables or other kinds of intermediate values for readability?

File: PathPlanning/DStarLite/d_star_lite.py [hunk 2]
```diff
@@ -62,32 +65,46 @@ def __init__(self, ox: list, oy: list):
...
+        y = np.array([node.y])
+        is_in_obstacles = (
+            (self.obstacles_xy[:, 0] == x)
+            & (self.obstacles_xy[:, 1] == y)
+        ).any()
+        is_in_detected_obstacles = False
+        if self.detected_obstacles_xy.shape[0] > 0:
+            is_in_detected_obstacles = (
+                (self.detected_obstacles_xy[:, 0] == x)
+                & (self.detected_obstacles_xy[:, 1] == y)
```
Comment (AtsushiSakai) [inline PathPlanning/DStarLite/d_star_lite.py:104]: Difficult to read each condition, is it possible to improve it?

File: PathPlanning/DStarLite/d_star_lite.py [hunk 3]
```diff
@@ -62,32 +65,46 @@ def __init__(self, ox: list, oy: list):
...
-        self.g = list()  # type: ignore
+        self.rhs = np.full(
+            (self.x_max, self.y_max),
+            float("inf")  # type: ignore
+        )
+        self.g = np.full(
+            (self.x_max, self.y_max),
+            float("inf")  # type: ignore
+        )
         self.detected_obstacles = list()  # type: ignore
```
Comment (AtsushiSakai) [inline PathPlanning/DStarLite/d_star_lite.py:84]: self.dtected_obstacles is still needed?

File: PathPlanning/DStarLite/d_star_lite.py [hunk 4]
```diff
@@ -62,32 +65,46 @@ def __init__(self, ox: list, oy: list):
...
         self.goal = Node(0, 0)
         self.U = list()  # type: ignore
         self.km = 0.0
         self.kold = 0.0
-        self.rhs = list()  # type: ignore
-        self.g = list()  # type: ignore
+        self.rhs = np.full(
+            (self.x_max, self.y_max),
+            float("inf")  # type: ignore
+        )
```
Comment (AtsushiSakai) [inline PathPlanning/DStarLite/d_star_lite.py:79]: I think introducing a new function to initialize a grid map is better like:

```
def init2dGrid():
       return np.full(self.x_max, self.y_max), float("inf")
```

File: PathPlanning/DStarLite/d_star_lite.py [hunk 5]
```diff
@@ -52,6 +53,8 @@ class DStarLite:
         Node(-1, -1, math.sqrt(2))
     ]
 
+    initialized = False
```
Comment (AtsushiSakai) [inline PathPlanning/DStarLite/d_star_lite.py:56]: This should be initizlied in the `__init__`?

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Thank you for your PR. I have some points to improve the readability of this code. PTAL.

File: PathPlanning/DStarLite/d_star_lite.py (hunk 1 above)
Comment (kadirhas) [reply PathPlanning/DStarLite/d_star_lite.py]: That is a great suggestion, thank you! I've updated this as you have suggested. Please take a look.

File: PathPlanning/DStarLite/d_star_lite.py (hunk 2 above)
Comment (kadirhas) [reply PathPlanning/DStarLite/d_star_lite.py]: I separated the conditions with boolean variables.

File: PathPlanning/DStarLite/d_star_lite.py (hunk 3 above)
Comment (kadirhas) [reply PathPlanning/DStarLite/d_star_lite.py:76]: It is still used in random object generation. I will change that function as well and test it.

File: PathPlanning/DStarLite/d_star_lite.py (hunk 4 above)
Comment (kadirhas) [reply PathPlanning/DStarLite/d_star_lite.py]: I used an already existing function called create_grid. Thanks for the improvement.

File: PathPlanning/DStarLite/d_star_lite.py (hunk 5 above)
Comment (kadirhas) [reply PathPlanning/DStarLite/d_star_lite.py]: Yes, I moved it, thank you

File: PathPlanning/DStarLite/d_star_lite.py (hunk 3 above)
Comment (kadirhas) [reply PathPlanning/DStarLite/d_star_lite.py]: I removed it, and while doing that fixed a bug about going out of range of the rhs.

**Review by AtsushiSakai — APPROVED** [review]
LGTM. Thank you for great PR!!

**Pull request merged**